Add support for encoding to DelimitedDataReader

neilculver opened the pull request:
Add an encoding to ensure the correct bytes are returned by `GetBytes()` by  and align `GetBytes()` behaviour with .NET framework implementations of `IDataReader`.

When returning the bytes representing text data we need to use an encoding to determine the correct byte representation. Casting directly from `char` to `byte` causes incorrect values for multi-bytes characters. Some behaviours of `GetBytes()`, such as reading data from an offset, have been update to match framework behaviour.

An encoding is now specified by:
- Passing a `StreamReader` to a new overload of `ParseReader()`. In this case the encoding on the `StreamReader` is used.
- Passing an explicit encoding along with a `TextReader` to `ParseReader()`.

Fixes #11

Commit:
Started adding support for an encoding to delimited data reader.

Commit:
Finished adding support for GetBytes with an encoding.

cgwrench reviewed (changes requested):
This mostly looks great @neilculver. However, I think that we should introduce additional overloads of `Parser.Parse()` and `Parser.ParseReader()`, rather than adding new optional parameters to the existing methods. This is because [introducing optional parameters to a shipped method can cause runtime issues](http://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/). Once this change has been made I think we can merge these changes in.

cgwrench committed:
Add parser method overloads

Remove optional encoding parameters from the `Parse` and
`ParseReader` methods in favour of method overloads, since adding
new optional parameters to existing methods can cause runtime issues.

cgwrench committed:
Add tests for encoding argument null checks

cgwrench committed:
Document parser encoding changes in CHANGELOG
